runner/src/main.rs

```diff
@@ -595,19 +653,267 @@ struct Example {
+                    application.construct_example_server_run_step(opt, example, run_clients)
+                }
+            }
+            _ => panic!(INVALID_APP_COMBO_ERR),
+        }
+    }
+}
+
+#[derive(serde::Deserialize, Debug)]
+#[serde(deny_unknown_fields)]
+struct ApplicationClassic {
     manifest: String,
     out: String,
     modules: HashMap<String, Target>,
 }
 
+impl ApplicationClassic {
+    fn construct_application_build_steps(&self, example_name: &str) -> Vec<Step> {
+        vec![
+            Step::Multiple {
+                name: "build wasm modules".to_string(),
+                steps: self
+                    .modules
+                    .iter()
+                    .map(|(name, target)| build_wasm_module(name, target, example_name))
+                    .collect(),
+            },
+            Step::Single {
+                name: "build application".to_string(),
+                command: build_application(&self),
+            },
+        ]
+    }
+
+    fn construct_example_server_run_step(
+        &self,
+        opt: &RunExamples,
+        example: &Example,
+        run_clients: Step,
+    ) -> Step {
+        let run_server = run_example_server(
+            &opt.build_server,
+            &example.server,
+            opt.server_additional_args.clone(),
+            &self.out,
+            &opt.permissions_file,
+        );
+
+        if opt.build_client.client_variant == NO_CLIENTS {
+            Step::Single {
+                name: "run server".to_string(),
+                command: run_server,
+            }
+        } else {
+            Step::WithBackground {
+                name: "background server".to_string(),
+                background: run_server,
+                foreground: Box::new(run_clients),
+            }
+        }
+    }
+}
+
+#[derive(serde::Deserialize, Debug)]
+#[serde(deny_unknown_fields)]
+struct ApplicationFunctions {
+    wasm_path: String,
+    target: Target,
+}
+
+impl ApplicationFunctions {
+    fn construct_application_build_steps(&self, example_name: &str) -> Vec<Step> {
+        vec![build_wasm_module(example_name, &self.target, example_name)]
+    }
+
+    fn construct_example_server_run_step(
+        &self,
+        opt: &RunFunctionsExamples,
+        example: &Example,
+        run_clients: Step,
+    ) -> Step {
+        let run_server = run_functions_example_server(&example.server, &self);
+
+        if opt.build_client.client_variant == NO_CLIENTS {
+            Step::Single {
+                name: "run server".to_string(),
+                command: run_server,
+            }
+        } else {
+            Step::WithBackground {
+                name: "background server".to_string(),
+                background: run_server,
+                foreground: Box::new(run_clients),
+            }
+        }
+    }
+}
+
+#[derive(serde::Deserialize, Debug)]
+#[serde(deny_unknown_fields)]
+struct ApplicationsList<T> {
+    applications: HashMap<String, T>,
```

Comment:
I guess this now results in an extra `applications` field everywhere, and having to forward methods like `is_empty` and `get`, let's think (in a separate PR) if it can be removed perhaps.

Reply:
I agree. Removed this class, but that resulted in some repetitions in the code. I'll try to fix that in a separate PR, together with other potential refactoring.